### PR TITLE
Removing support

### PR DESCRIPTION
**Pull request by ThomasWeiser**

### CODE REVIEW

**Comment by ThomasWeiser** [comment]
Implemented. Notes:

- Needs better UI. Currently user has to click on the indicator _SUPPORTING_ in proposal view, which is actually a button. We need something more discoverable and correctly labeled.
- There's a asymmetry here between adding and removing support at API level:
    - `POST /supports` responds with a complete `support` resource, informing the client about the new total number of supporting users on that proposal.
    - `DELETE /proposal/:id/support` has an empty response body (in conformance with JSON API / REST?). So the client just updates the support-count in its local model.

File: src/Api.elm [hunk 1]
```diff
@@ -256,13 +262,20 @@ createProposal proposalInput accessToken wrapMsg =
 
 supportProposal : String -> Bool -> String -> (Msg -> a) -> Cmd a
 supportProposal id newState accessToken wrapMsg =
```
Comment (oliverbarnes) [inline src/Api.elm]: how about also changing the function name to something like `toggleSupport`, since it does both support and support removal?

**Comment by oliverbarnes** [comment]
> - Needs better UI. Currently user has to click on the indicator _SUPPORTING_ in proposal view, which is actually a button. We need something more discoverable and correctly labeled.

Agreed. The button is deliberate, meant to be an indicator and actionable button at the same time, a toggle. Like a 👍 button... Perhaps rewording to `remove support` or `unsupport` (this sounds weird, though - haven't found this in a dictionary either) would make it clearer. Any other ideas? 

> - There's a asymmetry here between adding and removing support at API level:
    - `POST /supports` responds with a complete `support` resource, informing the client about the new total number of supporting users on that proposal.
    - `DELETE /proposal/:id/support` has an empty response body (in conformance with JSON API / REST?). So the client just updates the support-count in its local model.

Ahh, yes... this is an interesting problem. I've been mulling this over and haven't found a satisfying solution yet, and it might be a more general problem of api design for us, with implications for how we design interactions between the elm client and the api in the future. Do you have any ideas/opinions on this? 

With `POST /supports`, it's like you said - it is compliant with JSON API, and it makes sense to return the associated `proposal` resource included, with the updated `support-count` attribute. 

But when deleting, it feels weird to return an updated associated object, since it's associated to... nothing, now. And the JSON API response I'm used to is `204 No Content`, which means, well, nothing should be returned.

Talked to a fellow api dev about this, and he brought up the possibility of having a separate support-count endpoint - perhaps `GET /proposals/:id/support-count`, to be requested once `DELETE /proposals/:id/support` responds.

Also went back to the spec to double check on the accepted responses, and looking at the section [for updating **relationships**](http://jsonapi.org/format/1.1/#crud-updating-relationship-responses-200), which is the case now that we're using the `/proposals/:id/support/` scheme, it may be that actually returning updated resources might be compliant:

> If a server accepts an update but also changes the targeted relationship(s) in other ways than those specified by the request, it MUST return a 200 OK response. The response document MUST include a representation of the updated relationship(s).

Not sure what the payload format should be for this, though. Thinking it could be just the `included` object with the updated `proposal` resource, and this only with the updated `support-count` and `supported-by-me` attributes, in order to update the local `proposal` resource. Come to think of it, the previous solution doesn't take care of updating `supported-by-me`, either.

We could also just send another request to `GET /proposals/:id` once `DELETE /proposals/:id/support` responds, and get the updated `proposal` fields directly from that, reconciling with our local model.

Underlying all this is still my confusion about what to do about caching, local data layer, etc, and hearing recent talk about simplifying front-end/backend data exchange, and questioning of whether JSON-API is really appropriate for this.

**Review by oliverbarnes — COMMENTED** [review]
See previous comments. There's also the conflict with the recently merged 0.18 update

File: src/Api.elm (hunk 1 above)
Comment (ThomasWeiser) [reply src/Api.elm]: agreed

**Comment by ThomasWeiser** [comment]
Ok, PR is now updated to 0.18 too.

I'm not sure about the API design questions.

I don't think that the asymmetry regarding getting or not getting a response with updated data is a real problem for now. It just feels a little bit weird to have these asymmetric endpoints.

Alternatively `supported-by-me`could be handled purely as an attribute of `proposal`. Adding and removing support would be a `PATCH` [request](http://jsonapi.org/format/1.1/#crud-updating-resource-attributes). Server would [respond](http://jsonapi.org/format/1.1/#crud-updating-responses-200) with an updated `proposal` resource. But I don't know if this is any better.

Currently I don't have a clear picture yet regarding the details of data modelling at the API layer. Do you have a draft of the prospective entire API?

**Comment by oliverbarnes** [comment]
> Currently I don't have a clear picture yet regarding the details of data modelling at the API layer. Do you have a draft of the prospective entire API?

I don't, and I don't think we should have one at this point. I'd like to continue what we're doing, shaping the api as close as possible to front-end needs. Your `PATCH /proposals/:id` for changing the `supported-by-me` attribute goes in the right direction, I think: one request changing just one resource, without fussing with nested/included stuff. Nice and flat, mapping nicely to what we have displayed on the interface. `Proposal` - `Support` entity relationship is really a backend concern.

I agree the current solution is not a huge problem right now, but refactoring this could make our design decisions easier in the future, I'll try to submit something soon (perhaps with the Elm part as well ;) )

**Pull request merged**